### ISSUE TITLE
Feature: additionally publish uncompressed feather + gzip for JS support

### ISSUE THREAD

**Issue by larsyencken** [owid/etl#30]
We want to support both Python and JS catalog APIs, but current JS libraries don't seem to support compression of feather files. Still we want to use feather, and compressed too, to keep binary file sizes down and make it easy to work with.

@danyx23 suggests switching to uncompressed feather files as our default, and gzipping them. This should give us better compression and open up JS to using our catalog too.

## TODO

- [ ] Adapt `Table.read_feather()` and `Table.to_feather()` in `owid-catalog-py` to handle `.feather.gz` files
- [ ] Adapt `Dataset.add()` to add in feather-gz format by default
- [ ] Migrate `catalog.feather` to feather-gz format
- [ ] Rebuild and resync the catalog
- [ ] Release new `owid-catalog` version with changes

**Comment by danyx23** [comment]
I think we should leave this issue open as I hope we'll get to deciding and implementing this in the Mar/Apr iteration

**Comment by larsyencken** [comment]
This seems very related to our ongoing API discussions and to the JS API project for GSoC, we should revisit it after our offsite.

**Comment by wcox13** [comment]
> A fly in the soup: reading gzipped feather files is way slower than reading native feather compression. I'd suggest we generate both formats on the fly when we publish data, and JS simply uses the .feather.gz files.

@larsyencken @Marigold Given that we want to continue storing the existing format, does this ticket actually require a migration of existing code / data?

In more detail, here is an initial approach I'm thinking of after looking at the code a bit:

1. In `Table#to_feather`, also write an uncompressed + gzipped feather file, in addition to the existing feather with native compression
2. Add a `Table#read_feather_gz` to read it, factoring out common logic with existing `Table#read_feather` (or, add an additional flag argument to indicate whether to read the gzipped version)
3. In `CatalogSeries#load`, add an addition branch: `if self.format == "feather_gz": return Table.read_feather_gz(uri)`
4. When using the API from JS, find a way to pass this format indicator of `feather_gz` (I'm less clear on exactly the code change, but we can get to that if I'm on the right track here)

(I see there are additional usages of `read_feather` in `datasets` as well as `catalogs`, but assuming we could follow a similar approach)

**Comment by Marigold** [comment]
@wcox13 thanks for bringing life to this thread! These are good points. I'm actually thinking that if we store both formats then we might not even need reading from python at all (since reading feather files with current `zstd` compression should be more efficient). It'd be cool to create a simple benchmark for gzip & zstd comparing compression size and read speed. If we find that zstd is better in both then we only do writing with gzip.

**Comment by wcox13** [comment]
@Marigold Agreed! I can do some benchmarking. I'll use the df example you used since it seems to be a decent size, unless there is a better suited example: `https://owid-catalog.nyc3.digitaloceanspaces.com/garden/ggdc/2020-10-01/ggdc_maddison/maddison_gdp.feather`

**Comment by wcox13** [comment]
Alright, the benchmarking ended up being a decent chunk of code, so I'm posting the code + analysis at https://github.com/owid/etl/pull/183. Probably worth a look to make sure I'm measuring the right thing.

TLDR: compressed file sizes are similar, but it's about 6x slower to read using gzip (as Lars said).

With that confirmed, I think it's best to just write the additional format for now, to be handled later by JS @Marigold

**Comment by danyx23** [comment]
@wcox13 this look good already! I have a related request - could you add parquet file writing as well and add a benchmark for it? I think parquet does gzip compression by default and that should be fine.

The reason this is interesting is that we are considering using [duckdb](https://duckdb.org/) for some scenarios (including their version that is compiled to WASM and runs in the browser) and that seems to work with parquet files just fine which should be smaller still than compressed feather files. Thanks for your effort!

**Comment by Marigold** [comment]
Thanks @wcox13! I'd maybe use a larger dataset (couple MBs) since Maddison is like the smallest one that we currently have and larger ETL files are usually the bottleneck. I don't think it'd be much different though, `zstd` is _really_ fast according to many benchmarks.

I like this idea Daniel. Having a benchmark with all those formats could help our intuition when deciding what to go for. For context, I've done [some similar benchmarks](https://github.com/owid/notebooks/blob/main/MojmirVinkler/2022-04-27%20Wide%20vs%20long%20format.ipynb) on this in the past. The use case was slightly different, but could serve as a reference.

@wcox13 as for duckdb, I encourage you to try it! It's an amazing piece of technology and having someone with deep expertise could be invaluable.

**Comment by wcox13** [comment]
Thanks for the replies guys! For some reason just saw @danyx23 , updating my email settings...

> @wcox13 this look good already! I have a related request - could you add parquet file writing as well and add a benchmark for it? I think parquet does gzip compression by default and that should be fine.

For sure, will add a section for parquet to measure size/read speed

> I'd maybe use a larger dataset (couple MBs) since Maddison is like the smallest one that we currently have and larger ETL files are usually the bottleneck

Can do - I actually did try appending this dataset to itself to make it bigger. Interesting side point, `ztsd` was smart enough to catch the repeating data, and the compressed file was nearly the same size (though of course this threw off benchmarking since `gz` was not smart enough to do that). I'm not sure if we have a lot of repeating data like that (I would guess not), but if we do that's another point in favor of `zstd`

> as for duckdb, I encourage you to try it! It's an amazing piece of technology and having someone with deep expertise could be invaluable.

Sounds good - not familiar so it will be good to learn about. I'll play with it and try to see what the advantages are

**Comment by wcox13** [comment]
@danyx23 it looks like the default is actually "snappy" compression when writing a dataframe to parquet, is that OK or do you want to see the numbers specifically for gzip? Based on prelim googling snappy seems fine, but wanted to check

<img width="463" alt="Screen Shot 2022-05-12 at 2 52 26 PM" src="https://user-images.githubusercontent.com/6326346/168157140-1874633f-534a-49b6-bbfe-102c22d40be2.png">

**Comment by wcox13** [comment]
Actually, I just quickly compared the different compression types for Parquet. Compared to `gzip` or `brotli`, `snappy` has slightly higher file size but significantly better read performance. Also, they all have much higher file size and worse read performance than `zstd` for feather. Some numbers:

```
Dataframe len: 21854

Size of zstd feather: 186,626 bytes
Size of gzipped raw feather: 182,668 bytes
Size of parquet (snappy compression): 278,105 bytes

Average read time (zstd): 0.00194s
Average read time (uncompressed + gz): 0.01214s
Average read time (parquet + snappy): 0.00348s
```

**Comment by wcox13** [comment]
Also @Marigold, confirmed ratios are similar with bigger test file:

```
Dataframe len: 185444

Size of zstd feather: 8,777,546 bytes
Size of gzipped raw feather: 8,110,942 bytes
Size of parquet (snappy compression): 12,132,613 bytes

Average read time (zstd): 0.05284s
Average read time (uncompressed + gz): 0.87941s
Average read time (parquet + snappy): 0.09321s
```

**Comment by Marigold** [comment]
Thanks Billy, this will be helpful for the upcoming API project.

> it looks like the default is actually "snappy" compression when writing a dataframe to parquet, is that OK or do you want to see the numbers specifically for gzip? Based on prelim googling snappy seems fine, but wanted to check

snappy is a standard for big data tools like Spark or others, but we need to decompress it with javascript in the browser and **neither snappy nor zstd decompression** is supported (also the final size matters more than reading speed for us as transferring data over network is gonna be the bottleneck). That's why we have to do gzip or brotli which are the only ones supported.

This might drag for a while, so my suggestion is to add `gzip` compression to [Table#to_feather](https://github.com/owid/owid-catalog-py/blob/master/owid/catalog/tables.py#L87) (AFAIK it doesn't work out of the box) for now and then think about the best way how to publish this to S3.

The most straightforward is probably to add flag `--gzip` to `publish` command that would gzip the file on the fly (i.e. load it in memory as `Table`, call `to_feather(..., compression="gzip")` and upload it to S3 alongside the original file. So we'd end up with two files in the same folder, e.g. `wb_income_group.feather` and `wb_income_group.feather.gzip`. Syncing files to S3 is "smart" and [deletes files](https://github.com/owid/etl/blob/27913914992a51c02357495f8a2f82f652481ddd/etl/publish.py#L125) that are not in the catalog, so we'd have to make changes to that as well. I'm thinking that this publishing part is perhaps too complex and not worth doing yet as we don't have a clear idea how the API will work. So don't worry about this if you don't feel like it.

**Comment by wcox13** [comment]
> This might drag for a while, so my suggestion is to add gzip compression to [Table#to_feather](https://github.com/owid/owid-catalog-py/blob/master/owid/catalog/tables.py#L87) (AFAIK it doesn't work out of the box) for now and then think about the best way how to publish this to S3.

If we aren't close to supporting this on the JS end, should we even be adding the code to write the additional `gz` file until we do? It seems like it will be unnecessary complexity in the meantime.

> The most straightforward is probably to add flag --gzip to publish command that would gzip the file on the fly (i.e. load it in memory as Table, call to_feather(..., compression="gzip") and upload it to S3 alongside the original file

I'm not sure I fully understand this since I haven't looked at the publishing logic, but right now I'm just adding the following to `Table#to_feather`:

```
        # Since JS doesn't support native feather compression, we write both
        # 1) An uncompressed feather, which is then gzipped (for JS use)
        df.to_feather(path, compression=None, **kwargs)
        with open(path, 'rb') as raw:
            with gzip.open(path + '.gz', 'wb') as gzipped:
                shutil.copyfileobj(raw, gzipped)
        # 2) A feather file with native compression
        df.to_feather(path, compression=compression, **kwargs)
```

Is there a reason that approach doesn't work?

**Comment by Marigold** [comment]
> If we aren't close to supporting this on the JS end, should we even be adding the code to write the additional gz file until we do? It seems like it will be unnecessary complexity in the meantime.

I guess you're right, let's park this for now and keep it in the back of our minds.

> Is there a reason that approach doesn't work?

The function works fine, but a few things further downstream need to be changed to support two files (like syncing to S3).

**Comment by wcox13** [comment]
Makes sense 👍 I'll close the PR for now

**Comment by larsyencken** [comment]
Closing as "wontfix"
